runtime/gc_structs/ValueTypesIterator.hpp

```diff
@@ -0,0 +1,96 @@
+
+/*******************************************************************************
+ * Copyright (c) 1991, 2019 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at https://www.eclipse.org/legal/epl-2.0/
+ * or the Apache License, Version 2.0 which accompanies this distribution and
+ * is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following
+ * Secondary Licenses when the conditions for such availability set
+ * forth in the Eclipse Public License, v. 2.0 are satisfied: GNU
+ * General Public License, version 2 with the GNU Classpath
+ * Exception [1] and GNU General Public License, version 2 with the
+ * OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+/**
+ * @file
+ * @ingroup GC_Structs
+ */
+
+#if !defined(VALUETYPESITERATOR_HPP_)
+#define VALUETYPESITERATOR_HPP_
+
+#include "j9.h"
+#include "j9cfg.h"
+#include "modron.h"
+
+/**
+ * Iterate over all ValueType slots in a class which contain an object reference.
+ * @ingroup GC_Structs
+ */
+class GC_ValueTypesIterator
+{
+private:
+	/**
+	 * State constants representing the current stage of the iteration process
+	 * @anchor ValueTypeIteratorState
+	 */
+	typedef enum {
+		valuetypesiterator_state_start = 0,
+		valuetypesiterator_state_default_value,
+		valuetypesiterator_state_end
+	} ValueTypesState;
+	ValueTypesState _state;
+	
+	J9Class *_clazzPtr;
+	
+public:
+	GC_ValueTypesIterator(J9Class *clazz)
+		: _state(valuetypesiterator_state_start)
+		, _clazzPtr(clazz)
+	{};
+
+	/**
+	 * Fetch the next slot in the class.
+	 * Note that the pointer is volatile. In concurrent applications the mutator may 
+	 * change the value in the slot while iteration is in progress.
+	 * @return the address of the default value if the class is valuetype and is not array
+	 * @return NULL if not valuetype or is an array
+	 * or if the class has already been scanned
+	 */
+	j9object_t * nextSlot()
+	{	
+		/**
+		 * There is not a missed break.
+		 * The switch cases intentionally follow through
+		 * to the next state.
+		 */
+		switch(_state) {
+		case valuetypesiterator_state_start:
```

Comment:
Looks like you don't need special `valuetypesiterator_state_start` state. I would suggest to remove it and correspondent case in switch and set `_state = valuetypesiterator_state_end` in `valuetypesiterator_state_default_value` case